webgoat-lessons/challenge/src/main/java/org/owasp/webgoat/challenges/challenge7/Assignment7.java

```diff
@@ -0,0 +1,82 @@
+package org.owasp.webgoat.challenges.challenge7;
+
+import lombok.extern.slf4j.Slf4j;
+import org.owasp.webgoat.assignments.AssignmentEndpoint;
+import org.owasp.webgoat.assignments.AttackResult;
+import org.owasp.webgoat.challenges.Email;
+import org.owasp.webgoat.challenges.SolutionConstants;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.StringUtils;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.client.RestTemplate;
+
+import javax.servlet.http.HttpServletRequest;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.LocalDateTime;
+
+import static org.owasp.webgoat.challenges.Flag.FLAGS;
+
+/**
+ * @author nbaars
+ * @since 4/8/17.
+ */
+@RestController
+@Slf4j
+public class Assignment7 extends AssignmentEndpoint {
+
+    private static final String TEMPLATE = "Hi, you requested a password reset link, please use this "
+            + "<a target='_blank' href='%s:8080/WebGoat/challenge/7/reset-password/%s'>link</a> to reset your password."
+            + "\n \n\n"
+            + "If you did not request this password change you can ignore this message."
+            + "\n"
+            + "If you have any comments or questions, please do not hesitate to reach us at support@webgoat-cloud.org"
+            + "\n\n"
+            + "Kind regards, \nTeam WebGoat";
+
+    @Autowired
+    private RestTemplate restTemplate;
+    @Value("${webwolf.url.mail}")
+    private String webWolfMailURL;
+
+    @GetMapping("/challenge/7/reset-password/{link}")
+    public ResponseEntity<String> resetPassword(@PathVariable(value = "link") String link) {
+        if (link.equals(SolutionConstants.ADMIN_PASSWORD_LINK)) {
+            return ResponseEntity.accepted().body("<h1>Success!!</h1>"
+                    + "<img src='/WebGoat/images/hi-five-cat.jpg'>"
+                    + "<br/><br/>Here is your flag: " + "<b>" + FLAGS.get(7) + "</b>");
+        }
+        return ResponseEntity.status(HttpStatus.I_AM_A_TEAPOT).body("That is not the reset link for admin");
+    }
+
+    @PostMapping("/challenge/7")
+    @ResponseBody
+    public AttackResult sendPasswordResetLink(@RequestParam String email, HttpServletRequest request) throws URISyntaxException {
+        if (StringUtils.hasText(email)) {
+            String username = email.substring(0, email.indexOf("@"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

If the indexOf() search on line 61 does not find the substring and returns -1, the call to substring() on line 61 will throw IndexOutOfBoundsException. Note that the method is publicly accessible to external callers who might not be aware of the constraint on string email.